File: js/src/whalesong/stickerPack.js
import {
  ModelManager,
  CollectionManager
} from './common.js';
import {
  command
} from '../manager.js';
import {
  ValueError
} from '../errors.js';
import {
  ModelNotFound
} from './errors.js';


export class StickerManager extends ModelManager {

  static mapModel(item) {
    return Object.assign(ModelManager.mapModel(item), {
      _mediaObject: null
    });
  }


  @command
  async sendToChat({
    chatId,
    quotedMsgId
  }) {
    try {
      await this.model.downloadMedia();

      let chatManager = manager.getSubmanager('chats').getSubmanager(chatId);
    } catch (err) {
      throw new ModelNotFound(`Chat with ID "${chatId}" not found`);
    }

    if (quotedMsgId) {
      chatManager.model.composeQuotedMsg = chatManager.model.msgs.get(quotedMsgId);
    }

    return await chatManager._sendMessage(
      () => this.model.sendToChat(chatManager.model),
      (item) => item.type === 'sticker' && item.filehash === this.model.filehash
    );
  }
}

export class StickerCollectionManager extends CollectionManager {

  static getModelManagerClass() {
    return StickerManager;
  }

  @command
  async fetch() {
    await this.collection.fetch();
  }
}

export class StickerPackManager extends ModelManager {

  constructor(model) {
    super(model);

    this.addSubmanager('stickers', new StickerCollectionManager(this.model.stickers));
  }
}

export class StickerPackCollectionManager extends CollectionManager {

  static getModelManagerClass() {
    return StickerPackManager;
  }

  @command
  async fetchPage({
    page
  }) {

    let prIdx = this.collection.pageWithIndex(page);

    if (prIdx < page) {
      throw new ValueError(`Page ${page} does not exist`);
    }

    if (this.collection.pageFetchStates[prIdx] === 'SUCCESS') {
      return;
    }

    await this.collection.fetchAt(page);

    await this.collection._pageFetchPromises[prIdx];

    if (!this.collection.pageFetchStates[prIdx]) {
      throw new ValueError(`Page ${page} does not exist`);
    }
  }

  @command
  async fetchAllPages() {
    try {
      let page = 0;
      while (true) {
        await this.fetchPage({
          page: page
        });
        page++;
      }
    } catch (err) {};
  }

  @command
  async reset() {
    return this.collection.reset();
  }

  @command
  async getItemByName({
    name
  }) {
    let l = this.collection.where({
      name: name
    });

    if (l.length === 0) {
      throw new ModelNotFound(`Sticker pack with name "${name}" not found`);
    }

    return this.mapItem(l[0]);
  }
}

File: whalesong/managers/sticker_pack.py
from typing import Optional

from dirty_models import EnumField, StringIdField
from io import BytesIO

from . import BaseCollectionManager, BaseModelManager
from .message import MediaMixin, MessageTypes, download_media
from ..driver import BaseWhalesongDriver
from ..models import BaseModel
from ..results import Result


class Sticker(MediaMixin, BaseModel):
    type = EnumField(enum_class=MessageTypes, read_only=True, default=MessageTypes.STICKER)


class StickerPack(BaseModel):
    name = StringIdField()
    """
    Sticker pack name.
    """

    url = StringIdField()
    """
    Sticker pack url image.
    """


class StickerManager(BaseModelManager[Sticker]):
    """
    Sticker manager. It allows manage a sticker.
    """

    MODEL_CLASS = Sticker

    def send_to_chat(self, chat_id: str, quoted_msg_id: Optional[str] = None) -> Result[str]:
        """
        Send this sticker to a chat.

        :param chat_id: Chat indentifier where to send sticker.
        :param quoted_msg_id: Quoted message identifier.
        :return: Message identifier.
        """

        params = {'chatId': chat_id}

        if quoted_msg_id:
            params['quotedMsgId'] = quoted_msg_id

        return self._execute_command('sendToChat', params)

    async def download_image(self) -> BytesIO:
        """
        Download sticker's image file. It will decrypt image file using key on sticker object.

        :return: Image stream
        """
        model = await self.get_model()

        return await download_media(self._driver, model)


class StickerCollectionManager(BaseCollectionManager[StickerManager]):
    """
    Sticker collection manager. It allows manage sticker collection.
    """

    MODEL_MANAGER_CLASS = StickerManager

    def fetch(self) -> Result[None]:
        """
        Fetch all stickers. You must fetch stickers before try to us them.
        """
        return self._execute_command('fetch')


class StickerPackManager(BaseModelManager[StickerPack]):
    """
    Sticker pack manager. It allows manage a sticker pack.


    .. attribute:: stickers

        :class:`~whalesong.managers.sticker_pack.StickerCollectionManager`

        Sticker collection manager.
    """

    MODEL_CLASS = StickerPack

    def __init__(self, driver: BaseWhalesongDriver, manager_path: str = ''):
        super(StickerPackManager, self).__init__(driver=driver, manager_path=manager_path)

        self.add_submanager('stickers', StickerCollectionManager(
            driver=self._driver,
            manager_path=self._build_command('stickers')
        ))


class StickerPackCollectionManager(BaseCollectionManager[StickerPackManager]):
    """
    Sticker pack collection manager. It allows manage sticker pack collection.
    """

    MODEL_MANAGER_CLASS = StickerPackManager

    def fetch_page(self, page: int) -> Result:
        """
        Fetch a sticker pack page.
        """
        return self._execute_command('fetchPage', {'page': page})

    def fetch_all_pages(self) -> Result[bool]:
        """
        Fetch all sticker pack pages.
        """
        return self._execute_command('fetchAllPages')

    def reset(self) -> Result:
        """
        Reset sticker pack collection.
        """
        return self._execute_command('reset')

    def get_item_by_name(self, name: str) -> Result[StickerPack]:
        """
        Get sticker pack by name.

        :param name: Sticker pack name.
        :return: Sticker pack object.
        """
        return self._execute_command('getItemByName',
                                     {'name': name},
                                     result_class=self.get_item_result_class())
